File: Day 2: Arrays Part II/Merge Intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(),intervals.end());
        vector<vector<int>> res;
        int curr=0;
        res.push_back(intervals[0]);
        for(int i=1;i<intervals.size();i++){
            if(intervals[i][0]<=res.back()[1]){
                res.back()[0]=min(res.back()[0],intervals[i][0]);
                res.back()[1]=max(res.back()[1],intervals[i][1]);
            } else res.push_back(intervals[i]);
        }
        return res;
    }
};



/*
  O(NlogN)+O(N)-Time approach
  O(N)-Space approach
*/

File: Day 2: Arrays Part II/Java/Merge-Intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> res = new ArrayList<>();
        Arrays.sort(intervals, new Comparator<int[]>(){
            public int compare(int[] a, int[] b){
                return a[0]-b[0];
            }
        });
        int start = intervals[0][0];
        int end = intervals[0][1];
        for(int i=1;i<intervals.length;i++){
            int s = intervals[i][0], e = intervals[i][1];
            if(s>end){
                res.add(new int[]{start,end});
                start = s;
                end = e;
            } else {
                end = Math.max(end,e);
            }
        }
        res.add(new int[]{start,end});
        return res.toArray(new int[res.size()][]);
    }
}
